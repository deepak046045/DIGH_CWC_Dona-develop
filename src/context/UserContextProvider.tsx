'use client';

import React, {
  Dispatch,
  ReactNode,
  SetStateAction,
  createContext,
  useMemo,
  useState,
} from 'react';
import {
  NavigationOption,
  ungatedNavigationOptions,
} from '@/models/NavigationOptions';
import { Appointment } from '@/models/Appointment';
import { AlertType } from '@/models/AlertType';
import mockAppointments from '@/mocks/mockAppointments';
import { CarePartner } from '@/models/CarePartner';
import { mockCarePartners } from '@/mocks/mockCarePartners';

type UserContextDataType = {
  isLoggedIn: boolean;
  setIsLoggedIn: Dispatch<SetStateAction<boolean>>;
  activeView: NavigationOption;
  setActiveView: Dispatch<NavigationOption>;
  alert: AlertType;
  setAlert: Dispatch<SetStateAction<AlertType>>;
  appointments: Appointment[];
  setAppointments: Dispatch<SetStateAction<Appointment[]>>;
  carePartners: CarePartner[];
  setCarePartners: Dispatch<SetStateAction<CarePartner[]>>;
};

export const UserContext = createContext<UserContextDataType>({
  isLoggedIn: false,
  setIsLoggedIn: () => {},
  activeView: ungatedNavigationOptions[0],
  setActiveView: () => {},
  alert: { isOpen: false, message: '', level: '' },
  setAlert: () => {},
  appointments: mockAppointments,
  setAppointments: () => {},
  carePartners: mockCarePartners,
  setCarePartners: () => {},
});

type Props = {
  children: ReactNode;
};

export function UserContextProvider({ children }: Props) {
  const [isLoggedIn, setIsLoggedIn] = useState<boolean>(false);
  const [activeView, setActiveView] = useState<NavigationOption>(
    ungatedNavigationOptions[0]
  );
  const [alert, setAlert] = useState<AlertType>({
    isOpen: false,
    message: '',
    level: '',
  });
  const [appointments, setAppointments] =
    useState<Appointment[]>(mockAppointments);
  const [carePartners, setCarePartners] =
    useState<CarePartner[]>(mockCarePartners);

  const values = useMemo(
    () => ({
      isLoggedIn,
      setIsLoggedIn,
      activeView,
      setActiveView,
      alert,
      setAlert,
      appointments,
      setAppointments,
      carePartners,
      setCarePartners,
    }),
    [isLoggedIn, activeView, alert, appointments, carePartners]
  );

  return <UserContext.Provider value={values}>{children}</UserContext.Provider>;
}
