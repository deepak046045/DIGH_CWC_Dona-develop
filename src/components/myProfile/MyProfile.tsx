'use client';

import React, { useContext, useEffect } from 'react';
import Image from 'next/image';
import formatPhoneNumber from '@/utils/formatPhoneNumber';
import { useRouter } from 'next/navigation';
import {
  LdsButton,
  LdsModal,
  useLdsModal,
  LdsHorizontalRule,
} from '@elilillyco/ux-lds-react';
import modal from '@/scss/themes/modal.module.scss';
import buttons from '@/scss/themes/buttons.module.scss';
import { UserContext } from '@/context/UserContextProvider';
import mockUserData from '@/mocks/mockUserData.json';
import { ungatedNavigationOptions } from '@/models/NavigationOptions';
import { LdsModalType } from '@/models/LdsModal';
import styles from './MyProfile.module.scss';

export default function MyProfile() {
  // eslint-disable-next-line @typescript-eslint/no-unsafe-call
  const { isModalOpen, setIsModalOpen } = useLdsModal() as LdsModalType;

  const router = useRouter();
  const { isLoggedIn, setIsLoggedIn, setActiveView, setAlert } =
    useContext(UserContext);

  const handleCancel = () => {
    setIsModalOpen(false);
  };
  const handleDelete = () => {
    setIsModalOpen(false);
    setIsLoggedIn(false);
    setActiveView(ungatedNavigationOptions[0]);
    setAlert({
      isOpen: true,
      message: 'Your account has been successfully deleted.',
      level: 'success',
    });
    router.push('/');
  };
  const handleShowDialogBox = () => {
    setIsModalOpen(true);
  };

  useEffect(() => {
    if (!isLoggedIn) {
      router.push('/');
    }
  }, [isLoggedIn]);

  return (
    <div className={styles.container}>
      <div className={styles.title}>My profile</div>
      <div className={styles.mainContainer}>
        <div className={styles.profileIconContainer}>
          <Image
            src="/icons/profile-icon.svg"
            alt="ProfileIcon"
            width="29"
            height="29"
          />
          <div
            className={styles.profileIconDescription}
          >{`${mockUserData.firstName} ${mockUserData.lastName}`}</div>
        </div>
        <div className={styles.detailsContainer}>
          <div className={styles.detailsHeader}>FIRST NAME</div>
          <div className={styles.detailsValue}>{mockUserData.firstName}</div>
          <LdsHorizontalRule color="#ADB3B8" />
        </div>
        <div className={styles.detailsContainer}>
          <div className={styles.detailsHeader}>LAST NAME</div>
          <div className={styles.detailsValue}>{mockUserData.lastName}</div>
          <LdsHorizontalRule color="#ADB3B8" />
        </div>
        <div className={styles.detailsContainer}>
          <div className={styles.detailsHeader}>DATE OF BIRTH</div>
          <div className={styles.detailsValue}>{mockUserData.dateOfBirth}</div>
          <LdsHorizontalRule color="#ADB3B8" />
        </div>
        <div className={styles.detailsContainer}>
          <div className={styles.detailsHeader}>EMAIL ADDRESS</div>
          <div className={styles.detailsValue}>{mockUserData.email}</div>
          <LdsHorizontalRule color="#ADB3B8" />
        </div>
        <div className={styles.detailsContainer}>
          <div className={styles.detailsHeader}>MOBILE PHONE NUMBER</div>
          <div className={styles.detailsValue}>
            {formatPhoneNumber(mockUserData.mobilePhoneNumber)}
          </div>
          <LdsHorizontalRule color="#ADB3B8" />
        </div>
        <div className={styles.deleteContainer}>
          <LdsButton
            clearDefaultClasses
            classes={styles.deleteIconButton}
            onClick={handleShowDialogBox}
          >
            <Image
              src="/icons/delete-icon.svg"
              alt="DeleteIcon"
              width="19"
              height="19"
            />
            <div className={styles.deleteButtonLink}>Delete Account</div>
          </LdsButton>
          <div className={styles.deleteDescription}>
            Deleting your account only unregisters you from [sitename]. It does
            not affect your infusions. To stop taking Kisunla, talk to your
            doctor.
          </div>
        </div>
      </div>

      <div className={`${modal.primary}`}>
        <LdsModal
          modalId="confirmation-modal"
          open={isModalOpen}
          setModalOpen={setIsModalOpen}
        >
          <h2 className={modal.modalTitle}>Your account will be deleted</h2>
          <p className={modal.modalContent}>
            You and your care partners will lose access to your treatment plan,
            appointments and appointment reminders.
          </p>
          <div className={modal.modalActions}>
            <LdsButton
              classes={`${buttons.secondary}`}
              onClick={handleCancel}
              data-testid="cancel-btn"
            >
              Cancel
            </LdsButton>
            <LdsButton
              classes={`${buttons.primary}`}
              onClick={handleDelete}
              data-testid="yes-btn"
            >
              Yes, delete
            </LdsButton>
          </div>
        </LdsModal>
      </div>
    </div>
  );
}
