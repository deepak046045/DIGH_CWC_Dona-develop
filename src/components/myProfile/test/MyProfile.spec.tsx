import React from 'react';
import { useRouter } from 'next/navigation';
import { UserContext } from '@/context/UserContextProvider';
import {
  render,
  fireEvent,
  screen,
  waitFor,
  act,
} from '@testing-library/react';
import MyProfile from '../MyProfile';
import '@testing-library/jest-dom';

jest.mock('next/navigation', () => ({
  useRouter: jest.fn(() => ({ push: jest.fn() })),
}));
jest.mock('crypto', () => ({ randomUUID: () => '123456789' }));

describe('My Profile Component', () => {
  it('should render my profile view', () => {
    const { container } = render(<MyProfile />);
    expect(container).toMatchSnapshot();
  });
});

describe('when delete account link is clicked', () => {
  it('confirmation modal should be opened', () => {
    render(<MyProfile />);
    const deleteAccountBtn = screen.getByText('Delete Account');
    fireEvent.click(deleteAccountBtn);
    const modalTitle = screen.getByText('Your account will be deleted');
    expect(modalTitle).toBeTruthy();
  });
});

describe('when delete confirmation modal cancel button clicked', () => {
  it('should close the confirmation modal', async () => {
    const { container } = render(<MyProfile />);

    const deleteAccountBtn = screen.getByText('Delete Account');
    expect(container.querySelector('.open')).not.toBeInTheDocument();

    fireEvent.click(deleteAccountBtn);
    expect(container.querySelector('.open')).toBeInTheDocument();

    const cancelBtn = screen.getByTestId('cancel-btn');
    fireEvent.click(cancelBtn);

    await waitFor(() =>
      expect(container.querySelector('.open')).not.toBeInTheDocument()
    );
  });
});

describe('when delete confirmation modal yes button clicked', () => {
  it('should close the confirmation modal and redirect to the home page', async () => {
    const pushMock = jest.fn();
    (useRouter as jest.Mock).mockReturnValue({
      query: {},
      push: pushMock,
    });
    const { container } = render(<MyProfile />);
    const deleteAccountBtn = screen.getByText('Delete Account');
    expect(container.querySelector('.open')).not.toBeInTheDocument();

    fireEvent.click(deleteAccountBtn);
    expect(container.querySelector('.open')).toBeInTheDocument();

    const yesBtn = screen.getByText('Yes, delete');
    fireEvent.click(yesBtn);

    await waitFor(() => {
      expect(container.querySelector('.open')).not.toBeInTheDocument();
      expect(pushMock).toBeCalledWith('/');
    });
  });
});

describe('when isLoggedIn swaps to false', () => {
  it('should push "/" to the router', () => {
    const pushMock = jest.fn();
    (useRouter as jest.Mock).mockReturnValue({
      push: pushMock,
    });

    const { rerender } = render(
      <UserContext.Provider value={{ isLoggedIn: true }}>
        <MyProfile />
      </UserContext.Provider>
    );

    act(() => {
      rerender(
        <UserContext.Provider value={{ isLoggedIn: false }}>
          <MyProfile />
        </UserContext.Provider>
      );
    });

    expect(pushMock).toHaveBeenCalledWith('/');
  });
});
