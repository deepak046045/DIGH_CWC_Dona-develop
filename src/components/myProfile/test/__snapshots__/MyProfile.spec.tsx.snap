// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`My Profile Component should render my profile view 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="title"
    >
      My profile
    </div>
    <div
      class="mainContainer"
    >
      <div
        class="profileIconContainer"
      >
        <img
          alt="ProfileIcon"
          data-nimg="1"
          decoding="async"
          height="29"
          loading="lazy"
          src="/icons/profile-icon.svg"
          style="color: transparent;"
          width="29"
        />
        <div
          class="profileIconDescription"
        >
          Elizabeth Brown
        </div>
      </div>
      <div
        class="detailsContainer"
      >
        <div
          class="detailsHeader"
        >
          FIRST NAME
        </div>
        <div
          class="detailsValue"
        >
          Elizabeth
        </div>
        <lds-hr
          color="#ADB3B8"
        />
      </div>
      <div
        class="detailsContainer"
      >
        <div
          class="detailsHeader"
        >
          LAST NAME
        </div>
        <div
          class="detailsValue"
        >
          Brown
        </div>
        <lds-hr
          color="#ADB3B8"
        />
      </div>
      <div
        class="detailsContainer"
      >
        <div
          class="detailsHeader"
        >
          DATE OF BIRTH
        </div>
        <div
          class="detailsValue"
        >
          01/01/1990
        </div>
        <lds-hr
          color="#ADB3B8"
        />
      </div>
      <div
        class="detailsContainer"
      >
        <div
          class="detailsHeader"
        >
          EMAIL ADDRESS
        </div>
        <div
          class="detailsValue"
        >
          elizabethbrown@mail.com
        </div>
        <lds-hr
          color="#ADB3B8"
        />
      </div>
      <div
        class="detailsContainer"
      >
        <div
          class="detailsHeader"
        >
          MOBILE PHONE NUMBER
        </div>
        <div
          class="detailsValue"
        >
          (123) 456-7890
        </div>
        <lds-hr
          color="#ADB3B8"
        />
      </div>
      <div
        class="deleteContainer"
      >
        <button
          class="deleteIconButton"
          type="button"
        >
          <img
            alt="DeleteIcon"
            data-nimg="1"
            decoding="async"
            height="19"
            loading="lazy"
            src="/icons/delete-icon.svg"
            style="color: transparent;"
            width="19"
          />
          <div
            class="deleteButtonLink"
          >
            Delete Account
          </div>
        </button>
        <div
          class="deleteDescription"
        >
          Deleting your account only unregisters you from [sitename]. It does not affect your infusions. To stop taking Kisunla, talk to your doctor.
        </div>
      </div>
    </div>
    <div
      class="primary"
    >
      <div
        class="lds-modal"
        id="confirmation-modal"
      >
        <div
          class="wrapper"
        >
          <div
            class="row justify-content-center"
          >
            <div
              class="col-12 col-md-9"
            >
              <div
                aria-labelledby=""
                aria-modal="true"
                class="lds-modal-window"
                role="dialog"
              >
                <button
                  aria-label="close modal"
                  class="lds-button-clear-style lds-modal-close-btn focus icon"
                  type="button"
                >
                  <span
                    class="lds-button-icon-spacer"
                  >
                    Â 
                  </span>
                  <svg
                    aria-hidden="true"
                    aria-labelledby="lds-icon-X-title-123456789 lds-icon-X-description-123456789"
                    class="lds-icon icon-X inline button-icon"
                    focusable="false"
                    role="img"
                  >
                    <title
                      id="lds-icon-X-title-123456789"
                    >
                      X
                    </title>
                    <desc
                      id="lds-icon-X-description-123456789"
                    />
                    <svg
                      viewBox="0 0 256 256"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M205.66,194.34a8,8,0,0,1-11.32,11.32L128,139.31,61.66,205.66a8,8,0,0,1-11.32-11.32L116.69,128,50.34,61.66A8,8,0,0,1,61.66,50.34L128,116.69l66.34-66.35a8,8,0,0,1,11.32,11.32L139.31,128Z"
                      />
                    </svg>
                  </svg>
                </button>
                <div
                  class="lds-modal-body"
                >
                  <h2
                    class="modalTitle"
                  >
                    Your account will be deleted
                  </h2>
                  <p
                    class="modalContent"
                  >
                    You and your care partners will lose access to your treatment plan, appointments and appointment reminders.
                  </p>
                  <div
                    class="modalActions"
                  >
                    <button
                      class="lds-button secondary"
                      data-testid="cancel-btn"
                      type="button"
                    >
                      Cancel
                    </button>
                    <button
                      class="lds-button primary"
                      data-testid="yes-btn"
                      type="button"
                    >
                      Yes, delete
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
